### ISSUE TITLE
Eigenvalue computation changes in recent version

### ISSUE THREAD

**Issue by timlai4** [theislab/cellrank#349]
Dear Cellrank team,

Thank you for your tremendous help on my previous question https://github.com/theislab/cellrank/issues/297 and I'm sorry to trouble you again. 

I was previously working with v1.0.0rc0 or rc1 I believe where I ran the basic workflow successfully, then ran most of the advanced workflow successfully, up until the previous question. From the basic workflow, I saw that my sample had four terminal states. When I ran the advanced workflow with `backward = False` and computed the eigenvalues, I plotted them via `g.plot_spectrum(real_only = True)` and the eigengap in the picture separated out four eigenvalues. If I'm not mistaken, **the four eigenvalues should correspond to the four terminal states from the basic workflow and is the correct setting for the parameters `use = 4` in `plot_schur()` and `n_states = 4` in `compute_metastable_states()`**. Please let me know if I have conceptually misunderstood this. I inferred this from the fact that the workflow showed 3 eigenvalues past the eigengap and the rest of the parameters were set to 3. Also, the basic workflow recovered 3 terminal states (Alpha, Beta, Epsilon). 

![test](https://user-images.githubusercontent.com/46388504/91625298-343b4880-e974-11ea-84c3-f8d77a16790a.png)

However, now that I have upgraded to the newest version, v1.0.0rc6, I noticed that the eigengap only separates out a single eigenvalue. **I am wondering what was changed between the versions as far as computing the eigenvalues** that might result in such a significant decrease? Moreover, I am concerned because this sample I am analyzing does not have any knockdown or other biological significance that would indicate biologically only a single final state. 

![test2](https://user-images.githubusercontent.com/46388504/91625414-1fab8000-e975-11ea-8780-779c3d1e4407.png)

Finally, before I realized the number of eigenvalues decreased, I had still set `use = 4` and `n_states = 4` and it seemed the code would still run, produce figures, and states. I'm guessing if we ask for more eigenvalues beyond the eigengap, the program will still generate the statistics for them, but they may not be as "significant"?

Thank you again.

**Comment by Marius1311** [comment]
Hi @timlai4, between rc1 and rc6, we changed the way in which we compute the transition matrix. Before, we used scvelo's velocity graph whereas now, we compute this ourselves in the `VelocityKernel`. Try setting `softmax_scale=None` in `kernel.compute_transition_matrix` in the advanced workflow, and let me know whether this recovers the old spectrum (4 eigenvalues before the eigengap). 

@michalk8 , we should make `softmax_scale=None` the default value - changing this can have quite significant consequences.

**Comment by Marius1311** [comment]
Re your other question, the eigengap is an indication of what may be a "good" number of final states. However, you may choose any number you like. As you said correctly, choosing more states may just mean that some of them are not very significant of meaningful. More precisely, these will be less stable states, so it may include intermediate, metastable rather than final states.

**Comment by Marius1311** [comment]
Hi @timlai4 , any news on this?

**Comment by timlai4** [comment]
Thank you for the clarification. I added this to my transition matrix computation and the new eigenvalue picture looks like:
![test3](https://user-images.githubusercontent.com/46388504/91992508-38fc5580-ed02-11ea-98f4-dae50bbff515.png)

I have two eigenvalues before the eigengap now, but it's still not quite the same as the original picture. The eigengap is before the fourth eigenvalue and it seems a fifth eigenvalue is close to the magnitudes of the third and fourth as well. I noticed some of the previously real eigenvalues are colored as complex, but when plotting in the complex plane, I've noticed all the eigenvalues are basically on the real line, so I doubt this is significant beyond indicating that the computation has at least numerically changed.

**Comment by Marius1311** [comment]
thanks for trying this. Could you please post the code you use to compute the transition matrix here?

**Comment by Marius1311** [comment]
one more question, how many cells are we looking at?

**Comment by timlai4** [comment]
```{python}
scv.tl.velocity_graph(adata, mode_neighbors='connectivities', compute_uncertainties=True)
vk = VelocityKernel(adata)
vk.compute_transition_matrix(scale_by_variance = True, softmax_scale=None)
g = GPCCA(vk)
g.compute_schur(n_components=20)
g.plot_spectrum(real_only = True)
```
> one more question, how many cells are we looking at?
```{python}
>>> print(adata)
AnnData object with n_obs × n_vars = 3488 × 2999
```
Looks like 3,488 cells.

**Comment by Marius1311** [comment]
Hi @timlai4, let's try the following:
1. please update to the most recent version of cellrank
1. we simplified our interface a bit, among other changes, it does not matter any longer how you compute the velocity graph in scvelo becasue we internally re-compute it in cellrank (so you don't need to compute it at all in scvelo). Also, softmax scale is the default now. So simply run:

```pytb
vk = VelocityKernel(adata)
vk.compute_transition_matrix()
g = GPCCA(vk)
g.compute_schur(n_components=20)
g.plot_spectrum(real_only = True)
```

please, set the verbosity level to 4 (`cr.settings.verbosity = 4`) and post all output that the above commands give you here.

**Comment by Marius1311** [comment]
@timlai4 , I got it - I understand why you see a different spectrum now. Basically, we changed the way in which we propagate uncertainty into the transition probabilities. The old way was to do it via `scale_by_variance = True`. We got rid of this in favour for a more rigorous approach that propagates uncertainty using first principles, either through Monte Carlo simulations or though an analytical second order Taylor approximation. This is now controlled via the `mode` parameter, which can be `monte_carlo` for the former and `stochastic` for the latter. 

So in short: we made a backwards breaking change to the way in which we treat uncertainty in the velocity vectors. Sorry about that, but these things happen in early development. I can only reassure you that the new way to uncertainty is more rigorous than the old and the results are more trustworthy. Feel free to choose a greater number of sates than suggested by the eigengap - 4 or 5 looks reasonable for your data. 

@michalk8, this should clear things up.
